[GeoMechanicsApplication]  Move normal load logic to cpp

markelov208 opened the pull request:
**📝 Description**
- moved python class apply_normal_load_table_process to cpp

**🆕 Changelog**

- created cpp class ApplyNormalLoadTableProcess
- cleaned python script apply_normal_load_table_process .py
- moved free functions from apply_scalar_constraint_table_process.cpp to a new class FunctionsForParameters
- added ApplyNormalLoadTableProcess in dgeosettlement
- added ApplyNormalLoadTableProcess in add_custom_process_to_python

rfaasse committed:
Added new class for the normal load

avdg81 committed:
Removed an unused function and added an empty constructor

avdg81 committed:
[Incomplete] Started to move logic from Python to C++

Commit:
first version

avdg81 committed:
Fixed a few mistakes

- Corrected the return type of a copy assignment operator.
- Added a missing default destructor for class
  `ApplyNormalLoadTableProcess`.
- Added a missing `KRATOS_API(GEO_MECHANICS_APPLICATION)` (to export the
  class and make it available to the Python interface).
- Added a missing pointer definition.
- Removed an incorrect `return` statement from a Python script.
- Fixed an incorrect condition (we missed the comparison to 0).

Commit:
introduced functions in apply_normal_load_table_process

moved free functions from apply_scalar_constraint_table_process to a new class FunctionsForParameters
added apply_normal_load_table_process in dgeosettlement

Commit:
replaced some of Parameters operations with vector<string>

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/functions_for_parameters.cpp:
```diff
@@ -0,0 +1,62 @@
+        return result;
+    }
+
+    void FunctonsForParameters::AppendParameterNameIfExists(const std::string& rParameterName,
```
Minor comment about the styling: The functions in this file don't need to be indented here, inside namespaces we don't use indentation

markelov208 replied:
Thank you. It is removed now.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/functions_for_parameters.h:
```diff
@@ -0,0 +1,37 @@
+            static bool HasTableAttached(const Parameters& rSettings, int component);
+    };
+
+}
```
Also here, no indentation is needed after namespace (only for functions within the class itself)

markelov208 replied:
It is done.

Commit:
Merge branch 'master' into geo/move-normal-load-logic-to-cpp

# Conflicts:
#	applications/GeoMechanicsApplication/custom_processes/apply_scalar_constraint_table_process.cpp

Commit:
resolved merge conflicts

rfaasse reviewed (dismissed):
Nice clean and readable code! I only have 2 minor style related comments.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/functions_for_parameters.h:
To be consistent with the other files in the `custom_utilities` directory, I would suggest to rename this file to `parameters_utilities.h`. What do you think?

markelov208 replied:
Fully agreed, so the files are renamed etc.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/functions_for_parameters.h:
```diff
@@ -0,0 +1,37 @@
+
+namespace Kratos {
+
+    class FunctonsForParameters
```
And perhaps rename the class to `ParametersUtilities`? Otherwise, I would at least correct a typo: `FunctionsForParameters` (note the missing `i` in `Functons`).

markelov208 replied:
the class is renamed as yo usuggested.

avdg81 commented on applications/GeoMechanicsApplication/custom_processes/apply_normal_load_table_process.cpp:
```diff
@@ -0,0 +1,195 @@
+{  
+    auto normal_parameters = FunctonsForParameters::ExtractParameters(rProcessSettings,
+                                                                      NamesOfSettingsToCopy);
+    normal_parameters.AddValue("value", rProcessSettings["value"][0]);
```
I was just wandering whether the index here (`0`) also refers to the normal component of `value`...? If yes, then we might consider to define that index in a wider scope (e.g. as a class constant) and then replace those indices by the new constant. You have done that already in member functions `IsNormalComponentActive` and `IsTangentialComponentActive`, but perhaps we can use it here as well. What do you think?

markelov208 replied:
Indeed 0 and 1 are used in normal and tangential components. The constants are used now instead of the numbers.

avdg81 commented on applications/GeoMechanicsApplication/custom_processes/apply_normal_load_table_process.cpp:
```diff
@@ -0,0 +1,195 @@
+                                                               "second_reference_coordinate",
+                                                               "specific_weight"});
+    if (FunctonsForParameters::HasTableAttached(rProcessSettings)) {
+        NamesOfSettingsToCopy.insert(NamesOfSettingsToCopy.end(), { "table" });
```
Although correct, I would simplify this to:
```suggestion
        NamesOfSettingsToCopy.emplace_back("table");
```

markelov208 replied:
Thank you. It is modified now.

avdg81 commented on applications/GeoMechanicsApplication/custom_processes/apply_normal_load_table_process.cpp:
```diff
@@ -0,0 +1,195 @@
+                                                                "third_reference_coordinate",
+                                                                "specific_weight" });
+    if (FunctonsForParameters::HasTableAttached(rProcessSettings)) {
+        NamesOfSettingsToCopy.insert(NamesOfSettingsToCopy.end(), { "table" });
```
```suggestion
        NamesOfSettingsToCopy.emplace_back("table");
```

markelov208 replied:
As above.

avdg81 commented on applications/GeoMechanicsApplication/custom_processes/apply_normal_load_table_process.h:
```diff
@@ -0,0 +1,66 @@
+#include "processes/process.h"
+
+//#include <memory>
+//#include <vector>
```
Although strictly speaking these `#include`s could be left out (they are indirectly included through `processes/process.h`), I would insist to actually include them, since the class declares a data member (`mProcesses`) that requires them. We don't want to rely on accidental includes from other includes.

markelov208 replied:
restored.

avdg81 commented on applications/GeoMechanicsApplication/custom_processes/apply_normal_load_table_process.h:
```diff
@@ -0,0 +1,66 @@
+
+    void ExecuteInitialize() override;
+    void ExecuteInitializeSolutionStep() override;
+
```
Recently, we have added an override for member function `Info()` to class `ApplyScalarConstraintTableProcess`, which comes in really handy during debugging. I would suggest you to add a similar override here too.

markelov208 replied:
Thank you. Info() is added now.

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/functions_for_parameters.cpp:
```diff
@@ -0,0 +1,62 @@
+        const auto& table = rSettings["table"][component];
+        return std::any_of(table.begin(), table.end(),
+            [](const auto& value) {return value.GetInt() != 0; });
+    }
```
I doubt whether this is correct. My understanding is that `table` either has a single number or an array of numbers. Now this new overload checks one particular element from the array variant of `table`. By inspecting the original Python code, I don't have the impression that we might have a `table` which is an array, where each element contains an array in turn (which would require two indices to access a number). Therefore, I believe we can simplify this function body to:
```suggestion
        return rSettings["table"][component].GetInt() != 0;
```
Could you please verify whether my understanding is correct? And if yes, please correct the implementation as suggested above. Thanks!

markelov208 replied:
Thank you for pointing. I also found only [ "table"].GetInt() and ["table"][0/1/2].GetInt(). The check is removed now.

avdg81 reviewed (changes requested):
Overall, your changes look nice and clean. I have several minor suggestions and a few things that need to be checked. Could you please take care of that? Thanks!

Commit:
corrections in accordance with the review comments

- removed unnecessary indentation
- renamed functions_for_parameters to parameters_utilities
- used normal/tangential component instead of 0/1
- used emplace_back("table")
- reverted #include of <memory> and <vector>
- added Info()
- removed check if (rSettings["table"][component].IsNumber())

Commit:
- fixed indention

- removed KRATOS_ERROR_IF_NOT etc. for multi-component table

avdg81 committed:
Fixed some issues reported by SonarQube

- Made a few member functions `const`.
- Pass an object by reference to const rather than by value when it is
  not being modified.

rfaasse reviewed (approved):
Thanks for making the improvements, looks good!
